lib/netext/httpext/request_test.go

```diff
@@ -81,4 +85,27 @@ func TestMakeRequestError(t *testing.T) {
 		require.Error(t, err)
 		require.Equal(t, err.Error(), "unknown compressionType CompressionType(13)")
 	})
+
+	t.Run("invalid upgrade response", func(t *testing.T) {
+		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Add("Connection", "Upgrade")
+			w.Header().Add("Upgrade", "h2c")
+			w.WriteHeader(http.StatusSwitchingProtocols)
+		}))
```

Comment:
I'm not using `HTTPMultiBin` here to avoid a circular reference.

Sidenote: I know creating test servers at runtime like this is fairly inexpensive and often encouraged in Go testing, but this feels like a needlessly heavy end-to-end test, when a simple unit test would've been enough if `MakeRequest` worked with interfaces instead of concrete `http` APIs. Tools like [moq](https://github.com/matryer/moq) can help with generating test structs/mocks. Let me know if this can be avoided, and your thoughts on the interface approach.

Reply:
> I'm not using HTTPMultiBin here to avoid a circular reference.
This should be fixed. As we will need it at some point or another
In this particular part above you are basically doing what httpmultibin is doing ... just in a smaller scale. While I have no problems in the current situation  and as a quick solution for the fact httpmultibin is using netext(if I remember correctly), for something that is probably not needed - I am fine with it. And if this would be the only place where this is needed - it's fine.  But if we add more tests here (more accurately move them from the `http` js module here) we would probably need to use httpmultibin either way ;).

About the sidenote: My problem with mocking is that ... it tests your assumptions .. it doesn't test whether they are true. In the particular case, here it's probably fine, but in general I would say "No". Also if you go through most of the tests using httpmultibin .. if you need to set mocks for half of them it's probably going to be more work than the httpmultibin :( 
I would say that definitely there are places that we can optimize and there are use cases for mocks (database connections mainly ;)) but in general if the price is not high I prefer to use the real thing ;)

Reply:
> My problem with mocking is that ... it tests your assumptions .. it doesn't test whether they are true.

Agreed. Though typically mocking would be reserved for unit tests only. We should also have extensive integration and end-to-end tests, which we currently do. I suppose this distinction of the traditional [test pyramid](https://martinfowler.com/bliki/TestPyramid.html) is not that important in Go where the end-to-end approach in this case is very cheap. Still, having a clear separation of test type would be helpful in other ways. For a quick sanity check I'd rather run the full unit test suite in seconds, than a suite with tests of all type in minutes. I can run the full suite once before the last push, or wait for CI. But again, I'm not sure how important this is in Go, which has good test caching, from what I've seen.

Anyways, we can continue this discussion at another time/place. :)